### PR TITLE
formula_installer: accumulate inherited options

### PR DESCRIPTION
**Pull request by ilovezfs**
When a given dependency appears multiple times in a formula's dependency
tree, the inherited options for that dependency should accumulate rather
than being overwritten each time that dependency is considered by
expand_dependencies. In particular, this impacts "universal" since the
dependency should be built with universal unless all of its instances in
the dependency tree don't have "universal" as opposed to only if the last
instance considered has "universal."

See https://github.com/Homebrew/homebrew-core/issues/1604 for an example
of how the current code can cause build failure.

### CODE REVIEW

File: Library/Homebrew/formula_installer.rb [hunk 1]
```diff
@@ -344,7 +344,15 @@ def expand_dependencies(deps)
     inherited_options = {}
 
     expanded_deps = Dependency.expand(formula, deps) do |dependent, dep|
-      options = inherited_options[dep.name] = inherited_options_for(dep)
+      additional_inherited = inherited_options_for(dep)
+      unless additional_inherited.nil?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: `if additional_inherited` is simpler. I also personally hate `unless` .. `else` as I find the double-negatives hard to parse.

File: Library/Homebrew/formula_installer.rb [hunk 2]
```diff
@@ -344,7 +344,15 @@ def expand_dependencies(deps)
...
+        inherited_options[dep.name] = if inherited_options[dep.name].nil?
+          additional_inherited
+        else
+          inherited_options[dep.name] + additional_inherited
+        end
+      end
+      options = inherited_options[dep.name]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: I _think_ you can just do:
```ruby
options = inherited_options[dep.name] || []
options += additional_inherited if additional_inherited
options
```

Comment (ilovezfs) [inline Library/Homebrew/formula_installer.rb]: maybe
```
inherited_options[dep.name] ||= Options.new
options = inherited_options[dep.name] += inherited_options_for(dep)
```

Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: 👍 except you'll probably need to handle the `nil` `inherited_options_for(dep)` case but could do `Array(inherited_options_for(dep))`

Comment (ilovezfs) [inline Library/Homebrew/formula_installer.rb]: Right, but looks like `inherited_options_for` always returns at least `Options.new`, at least as currently implemented. But maybe best not to assume that implementation detail will always be true?

Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Nah, I think that's a fine assumption 👍

**Comment by MikeMcQuaid** [comment]
👍

**Comment by MikeMcQuaid** [comment]
(but I'd wait for a couple more reviews first)

File: Library/Homebrew/formula_installer.rb [hunk 3]
```diff
@@ -344,7 +344,8 @@ def expand_dependencies(deps)
     inherited_options = {}
 
     expanded_deps = Dependency.expand(formula, deps) do |dependent, dep|
-      options = inherited_options[dep.name] = inherited_options_for(dep)
+      inherited_options[dep.name] ||= Options.new
+      options = inherited_options[dep.name] += inherited_options_for(dep)
```
Comment (UniqMartin) [inline Library/Homebrew/formula_installer.rb]: I found the previous double assignment not exactly readable and it only got worse. How about the following?

```ruby
inherited_options[dep.name] ||= Options.new
inherited_options[dep.name] += inherited_options_for(dep)
options = inherited_options[dep.name]
```

And maybe just drop the assignment to `options` completely? If I'm not overlooking something, it is only used once in the following code and that instance could as well be replaced with `inherited_options[dep.name]`.

**Comment by UniqMartin** [comment]
Have added one code style comment, but otherwise the solution seems sound. (I haven't done extensive testing, though, and our dependency resolution code is—well—complicated.)

File: Library/Homebrew/formula_installer.rb (hunk 3 above)
Comment (ilovezfs) [inline Library/Homebrew/formula_installer.rb]: Yeah I'd been wondering why there was a separate copy if it was only then being passed by value in `dep.satisfied?(options)`.

**Comment by ilovezfs** [comment]
@UniqMartin PR refreshed with that change.

**Comment by ilovezfs** [comment]
@xu-cheng did you have a chance to review this PR?

File: Library/Homebrew/formula_installer.rb [hunk 4]
```diff
@@ -344,7 +344,8 @@ def expand_dependencies(deps)
     inherited_options = {}
 
     expanded_deps = Dependency.expand(formula, deps) do |dependent, dep|
-      options = inherited_options[dep.name] = inherited_options_for(dep)
+      inherited_options[dep.name] ||= Options.new
+      inherited_options[dep.name] += inherited_options_for(dep)
```
Comment (xu-cheng) [inline Library/Homebrew/formula_installer.rb]: A small nit. Maybe `<<` instead of `+=`.

**Comment by xu-cheng** [comment]
LGTM

File: Library/Homebrew/formula_installer.rb (hunk 4 above)
Comment (ilovezfs) [inline Library/Homebrew/formula_installer.rb]: Good idea.

**Comment by ilovezfs** [comment]
PR updated with the `<< ` change.

**Comment by UniqMartin** [comment]
> PR updated with the `<< ` change.

No! Unless I'm completely missing something, this will add the `Options` instance returned by the `inherited_options_for` method as a single element to the `Options` instance created in the `expand_dependencies` method. Definitely not the concatenation of arrays that should be here. (We want a flat array/list of options not an array of option arrays.)

And while we're iterating through various operators here, on second thought `|=` seems more appropriate, as we don't really want to have duplicate options, hence better use a set union.

**Comment by xu-cheng** [comment]
> No! 

Oops, I completely misread the code. @UniqMartin is right, it should be `|=` or `+=`.

**Comment by ilovezfs** [comment]
@UniqMartin Yes, I think you're right. Interestingly, although this last change causes my test of the underlying `brew install wine` failure to immediately blow up, the brew test framework didn't care at all. That's every so slightly discomforting.

**Comment by ilovezfs** [comment]
Here's how that failure presents:
```
iMac-TMP:local joe$ brew install wine
==> Installing dependencies for wine: xz, libpng, freetype, jpeg, libtool, pkg-config, libusb,
==> Installing wine dependency: xz
Error: undefined method `flag' for #<Options: []>
Please report this bug:
    https://git.io/brew-troubleshooting
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/set.rb:232:in `each_key'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/set.rb:232:in `each'
/usr/local/Library/Homebrew/options.rb:69:in `each'
/usr/local/Library/Homebrew/options.rb:102:in `map'
/usr/local/Library/Homebrew/options.rb:102:in `as_flags'
/usr/local/Library/Homebrew/formula_installer.rb:551:in `build_argv'
/usr/local/Library/Homebrew/formula_installer.rb:569:in `build'
/usr/local/Library/Homebrew/formula_installer.rb:236:in `install'
/usr/local/Library/Homebrew/formula_installer.rb:437:in `install_dependency'
/usr/local/Library/Homebrew/formula_installer.rb:388:in `block in install_dependencies'
/usr/local/Library/Homebrew/formula_installer.rb:388:in `each'
/usr/local/Library/Homebrew/formula_installer.rb:388:in `install_dependencies'
/usr/local/Library/Homebrew/formula_installer.rb:188:in `install'
/usr/local/Library/Homebrew/cmd/install.rb:275:in `install_formula'
/usr/local/Library/Homebrew/cmd/install.rb:149:in `block in install'
/usr/local/Library/Homebrew/cmd/install.rb:149:in `each'
/usr/local/Library/Homebrew/cmd/install.rb:149:in `install'
/usr/local/Library/brew.rb:87:in `<main>'
```

**Comment by ilovezfs** [comment]
`|=` ftw.

**Comment by ilovezfs** [comment]
Yeah, it appears the test framework doesn't have an inherited options test.

**Comment by UniqMartin** [comment]
> […], the brew test framework didn't care at all. That's every so slightly discomforting.

Sadly, we're quite a bit away from being able to rely on it. If it blows up, at least the likelihood is big that somethings broken. If it stays calm, this can really mean anything.

**Comment by ilovezfs** [comment]
>If it stays calm, this can really mean anything.

Indeed. Or as the aphorism goes, "absence of evidence is not evidence of absence."

**Comment by ilovezfs** [comment]
:shipit: ?

**Comment by UniqMartin** [comment]
I think the combination of `||=` and `|=` next to each other looks really weird, but oh well … From a technical standpoint: :shipit:

**Comment by MikeMcQuaid** [comment]
@ilovezfs As long as it works with your local testing 👍 🚢

**Comment by ilovezfs** [comment]
>I think the combination of ||= and |= next to each other looks really weird

LOL. I'm putting that in the `:uniqmartin-self-fix` bucket :)

**Comment by UniqMartin** [comment]
> LOL. I'm putting that in the `:uniqmartin-self-fix` bucket :)

:joy: Put this before the loop to initialize the hash:

```ruby
inherited_options = Hash.new { |hash, key| hash[key] = Options.new } 
```

And then only do this in the loop:

```ruby
inherited_options[dep.name] |= inherited_options_for(dep)```
```

Accessed elements are implicitly initialized and the problem is solved.

**Comment by UniqMartin** [comment]
Caveat for this suggestion: `inherited_options[<something>]` now never returns `nil` and one has to remember to use `Hash#key?`, `Hash#include?`, or `Hash#member?` to check whether an element exists in the hash. This can sometimes make quite a difference and break existing assumptions.

**Comment by ilovezfs** [comment]
I think I'm happy with this as is.

Also, under the hood
```
  def initialize(*args)
    @options = Set.new(*args)
  end

  def +(o)
    self.class.new(@options + o)
  end

 def |(o)
    self.class.new(@options | o)
  end
```
Since @options is a set, what's the difference between `+` and `|` other than appearance in this case.

**Comment by UniqMartin** [comment]
> Since `@options` is a set, what's the difference between `+` and `|` other than appearance in this case.

Technically, there's none. I didn't notice this implementation detail when I suggested to use `|=`, so my mention of duplicate options in the list/array is obviously not true as the underlying data structure is a set. I still think `|=` is subtly better at communicating what it is doing (set union) than `+=`.

**Comment by ilovezfs** [comment]
@UniqMartin also, I don't think the nil/non-nil assumptions have changed actually. As far as I can tell, the code without this PR leads to always-initialized too
```
  options = inherited_options[dep.name] = inherited_options_for(dep)
```
where inherited_options_for(dep) returns at minimum an Options.new. Or did I miss something?

**Comment by ilovezfs** [comment]
I guess you're saying `inherited_options = {}` leads to a difference outcome vs. `Hash.new { |hash, key| hash[key] = Options.new }` in the case where there are no deps at all?

**Comment by UniqMartin** [comment]
> I guess you're saying `inherited_options = {}` leads to a difference outcome vs. `Hash.new { |hash, key| hash[key] = Options.new }` in the case where there are no deps at all?

Not necessarily in this case, but yes that was what my caveats was referring to. That one has to be careful replacing `{}` with `Hash.new`-with-block. But let's just ship this as-is and be done!

**Comment by ilovezfs** [comment]
Ah yes, I see what you mean now. You were proposing to have it auto-initialize the value on access. I think that would probably be fine though, since that hash doesn't survive this method does it? The return value is
```
    expanded_deps.map { |dep| [dep, inherited_options[dep.name]] }
```

**Comment by ilovezfs** [comment]
@UniqMartin I refreshed the PR with your proposal. The two approaches seem to be approximately equally idiomatic in the code base, and it's one less line of code. Plus, it keeps the focus on what we care about in the iteration, which is the `|=`.

**Comment by UniqMartin** [comment]
Glad you like it. But more importantly: Great work at debugging and identifying the issue so quickly! 🎉

**Comment by ilovezfs** [comment]
Thanks :) Also three cheers for our awesome users who reported it so quickly, as it made identifying the commit in which the regression occurred straightforward with git bisect.
